lib/Geo/ShapeContainer.cpp

```diff
@@ -21,882 +21,652 @@
 
-#include <stddef.h>
-#include <algorithm>
-#include <array>
+#include "Geo/ShapeContainer.h"
+#include "Geo/GeoParams.h"
+#include "Geo/Ellipsoid.h"
+#include "Geo/Utils.h"
+#include "Geo/S2/S2MultiPointRegion.h"
+#include "Geo/S2/S2MultiPolyline.h"
+#include "Geo/karney/geodesic.h"
+#include "Basics/DownCast.h"
+#include "Basics/Exceptions.h"
 
 #include <s2/s1angle.h>
-#include <s2/s2cell.h>
-#include <s2/s2cell_id.h>
+#include <s2/s2latlng.h>
 #include <s2/s2latlng_rect.h>
-#include <s2/s2loop.h>
 #include <s2/s2point_region.h>
 #include <s2/s2polygon.h>
 #include <s2/s2polyline.h>
-#include <s2/s2region.h>
-#include <s2/s2region_coverer.h>
-#include <s2/util/math/vector.h>
 
-#include <velocypack/Slice.h>
+#include <cmath>
 
-#include "ShapeContainer.h"
+namespace arangodb::geo {
+namespace rect {
+// only used in legacy situations
 
-#include "Basics/Exceptions.h"
-#include "Basics/voc-errors.h"
-#include "Geo/Ellipsoid.h"
-#include "Geo/GeoParams.h"
-#include "Geo/S2/S2MultiPointRegion.h"
-#include "Geo/S2/S2MultiPolyline.h"
-#include "Geo/Shapes.h"
-#include "Geo/Utils.h"
-#include "Geo/karney/geodesic.h"
-#include "Logger/LogMacros.h"
-
-using namespace arangodb;
-using namespace arangodb::geo;
-
-namespace {
-
-S2Polygon latLngRectToPolygon(S2LatLngRect const* rect) {
-  // Construct polygon from rect:
-  auto first = rect->GetVertex(0).ToPoint();
-  std::array<S2Point, 5> v{first, rect->GetVertex(1).ToPoint(),
-                           rect->GetVertex(2).ToPoint(),
-                           rect->GetVertex(3).ToPoint(), first};
+static S2Polygon toPolygon(S2LatLngRect const& rect) {
+  std::array<S2Point, 4> v{
+      rect.GetVertex(0).ToPoint(), rect.GetVertex(1).ToPoint(),
+      rect.GetVertex(2).ToPoint(), rect.GetVertex(3).ToPoint()};
   auto loop = std::make_unique<S2Loop>(v, S2Debug::DISABLE);
   return S2Polygon{std::move(loop), S2Debug::DISABLE};
 }
 
-}  // namespace
-
-Result ShapeContainer::parseCoordinates(VPackSlice const& json, bool geoJson) {
-  if (!json.isArray() || json.length() < 2) {
-    return Result(TRI_ERROR_BAD_PARAMETER, "Invalid coordinate pair");
+static bool contains(S2LatLngRect const& rect, S2Polyline const& polyline) {
+  for (auto const& point : polyline.vertices_span()) {
+    if (!rect.Contains(point)) {
+      return false;
+    }
   }
+  return true;
+}
 
-  VPackSlice lat = json.at(size_t(geoJson));
-  VPackSlice lng = json.at(1 - size_t(geoJson));
-  if (!lat.isNumber() || !lng.isNumber()) {
-    return Result(TRI_ERROR_BAD_PARAMETER, "Invalid coordinate pair");
+static bool contains(S2Polygon const& polygon, S2LatLngRect const& rect) {
+  if (ADB_UNLIKELY(rect.is_point())) {
+    TRI_ASSERT(rect.lo().is_valid());
+    return polygon.Contains(rect.lo().ToPoint());
   }
-
-  resetCoordinates(lat.getNumber<double>(), lng.getNumber<double>());
-  return TRI_ERROR_NO_ERROR;
+  auto subregionBound = polygon.GetSubRegionBound();
```

Comment:
It's correct because
```
  // Since "bound_" is not exact, it is possible that a polygon A contains
  // another polygon B whose bounds are slightly larger.  "subregion_bound_"
  // has been expanded sufficiently to account for this error, i.e.
  // if A.Contains(B), then A.subregion_bound_.Contains(B.bound_).
  S2LatLngRect subregion_bound_;
```
```
bool S2Polygon::Contains(const S2Polygon& b) const {
  // It's worth checking bounding rectangles, since they are precomputed.
  // Note that the first bound has been expanded to account for possible
  // numerical errors in the second bound.
  if (!subregion_bound_.Contains(b.bound_)) {
    // simplifield
    return false;
  }
```
In general I just moved it from S2Polygon::Contains to not construct S2Polygon from S2LatLngRect

Reply:
ok. Understood.